src/core/qgsfield.cpp

```diff
@@ -589,6 +589,29 @@ bool QgsField::convertCompatible( QVariant &v, QString *errorMessage ) const
     return false;
   }
 
+  // Handle referenced geometries (e.g. from additional geometry fields)
+  if ( d->type == QVariant::String && v.type() == QVariant::UserType && v.canConvert<QgsReferencedGeometry>( ) )
```

Comment:
```suggestion
  if ( d->type == QVariant::String && v.userType() == QMetaType::type( "QgsReferencedGeometry" ) )
```

QVariant::canConvert will return true here for ANY qvariant type under qt 6. The meaning of canConvert has changed subtly from "is this a variant which stores a QgsReferencedGeometry" to "is it possible to construct a QgsReferencedGeometry from the value stored in the variant". And since QgsReferencedGeometry has a default constructor, the answer will ALWAYS be true.